mountpoint-s3/src/fs.rs

```diff
@@ -425,30 +428,92 @@ impl S3Personality {
             sse_kms_key_id,
+            checksum,
         }
     }
 
-    /// String representation of the SSE type as it is expected by S3 API
-    pub fn sse_type(&self) -> Option<String> {
-        self.sse_type.clone()
+    /// Computes the checksum of SSE settings by combining two strings containing the type and the key
+    /// Note, that this implementation yields the same result for Some("") and None, but we may safely
+    /// assume that it will never be called with an empty string as one of its parameters.
+    fn compute_checksum(sse_type: Option<&str>, sse_kms_key_id: Option<&str>) -> Crc32c {
+        let mut hasher: Hasher = Hasher::new();
+        if let Some(maybe_sse_type) = sse_type {
+            hasher.update(maybe_sse_type.as_bytes().as_ref());
+        }
+        if let Some(maybe_sse_kms_key_id) = sse_kms_key_id {
+            hasher.update(maybe_sse_kms_key_id.as_bytes().as_ref());
+        }
+        hasher.finalize()
     }
 
-    /// AWS KMS Key ID, if provided
-    pub fn key_id(&self) -> Option<String> {
-        self.sse_kms_key_id.clone()
+    /// Checks that SSE settings still match the checksum and returns the string representations of:
+    /// 1. the SSE type as it is expected by S3 API and;
+    /// 2. AWS KMS Key ID, if provided.
+    pub fn into_inner(self) -> Result<(Option<String>, Option<String>), IntegrityError> {
+        let computed = Self::compute_checksum(self.sse_type.as_deref(), self.sse_kms_key_id.as_deref());
+        if computed == self.checksum {
+            Ok((self.sse_type, self.sse_kms_key_id))
+        } else {
+            Err(IntegrityError::ChecksumMismatch(self.checksum, computed))
+        }
+    }
+
+    /// Checks that values provided as arguments to this function match the checksum stored in the object.
+    /// S3 will return some values for sse type and key even if they were not set on our side. We want to check
+    /// only the values which we set, thus response values are only added to checksum if they are present in this object.
+    pub fn verify_response(&self, sse_type: Option<&str>, sse_kms_key_id: Option<&str>) -> Result<(), IntegrityError> {
+        // we first validate in-memory values, since we are using them to compute a response's checksum
+        let computed = Self::compute_checksum(self.sse_type.as_deref(), self.sse_kms_key_id.as_deref());
+        if computed != self.checksum {
+            return Err(IntegrityError::ChecksumMismatch(self.checksum, computed));
+        }
+        // computing a response's checksum, adding values only if they were present in a request
+        let sse_type = self.sse_type.as_ref().and(sse_type);
+        let sse_kms_key_id = self.sse_kms_key_id.as_ref().and(sse_kms_key_id);
+        let computed = Self::compute_checksum(sse_type, sse_kms_key_id);
+        if computed == self.checksum {
+            Ok(())
+        } else {
+            Err(IntegrityError::ChecksumMismatch(self.checksum, computed))
+        }
```

Comment:
Using `.and` is a clever trick, maybe too clever? I'd rather revert to an explicit `if` block.

But more importantly, why compute and compare the checksum? Why not compare the actual strings like in the original `check_response_headers`?

Reply:
That's a good point, we can directly compare strings here, since we already validated that in-memory settings (`self.sse_type`, `self. sse_kms_key_id`) [match the checksum](https://github.com/awslabs/mountpoint-s3/pull/745/files/1a4cfc2e0656f203d916beddcbdd7453ce9c0321#diff-5397f952b2dcde7127c30d9a585c4e5a8491a0c70804e4b1e457c633637c63cdR489), so basically only compute the checksum once in `verify_response`. 

**We still need to compute the checksum once in this method though.** We will not do the check here, if an in-memory value is `None` (e.g. if `self.sse_kms_key_id == None`). Computing the checksum for in-memory values is the only way to ensure that the check won't be mistakenly skipped.